### ISSUE TITLE
[Bug] Circular Reference detection issue

### ISSUE THREAD

**Issue by ycherkes** [thomasgalliker/ObjectDumper#96]
```
        [Fact]
        public void ShouldDumpEnumerable_CorrectCircularReferenceDetection()
        {
            // Arrange 
            var obj = new { Prop = new { SomeInnerProp = "test_test_test" } };
            var enumerable = Enumerable.Range(0, 2).Select(_ => obj);

            // Act
            var dump = ObjectDumperCSharp.Dump(enumerable);

            // Assert
            this.testOutputHelper.WriteLine(dump);
            dump.Should().NotBeNull();
            dump.Should().NotContain("// Circular reference detected");
        }
```

**Comment by thomasgalliker** [comment]
Thanks, @ycherkes 
Would this be the correct output in your opinion?
```
var list = new List<dynamic>
{
    new { Prop = new { SomeInnerProp = "test_test_test" } },
    new { Prop = new { SomeInnerProp = "test_test_test" } }
};
```

**Comment by thomasgalliker** [comment]
@ycherkes NuGet pre-release https://www.nuget.org/packages/ObjectDumper.NET/3.3.23-pre contains the fix. Can you try this package and verify if you no longer experience the issue?

**Comment by ycherkes** [comment]
Hi @thomasgalliker,
Your fix works for previous test, but fails for this one:

        [Fact]
        public void ShouldDumpEnumerable_CorrectCircularReferenceDetection()
        {
            // Arrange 
            var example1 = new Example{ Name = "Name1"};
            var example2 = new Example { Name = "Name2", Reference = example1};
            var array = new[]{ example1, example2 };

            // Act
            var dump = ObjectDumperCSharp.Dump(array);

            // Assert
            this.testOutputHelper.WriteLine(dump);
            dump.Should().NotBeNull();
            dump.Should().NotContain("// Circular reference detected");
        }

        class Example
        {
            public string Name { get; set; }
            public object Reference { get; set; }
        }

**Comment by thomasgalliker** [comment]
I created a new branch #98 to solve this issue. I was looking into two json serializer frameworks to get an idea of how they handler circular reference problems. I think I found a solution which works pretty well. I pushed my latest changes to branch #98. If you @ycherkes could have a look at it, I would be very happy. There is no pre-release nuget available yet and the code needs some cleanup before I can merge it.
